src/gc.c

```diff
@@ -935,13 +935,29 @@ static jl_taggedvalue_t **sweep_page(jl_gc_pool_t *p, jl_gc_pagemeta_t *pg, jl_t
         int has_young = 0;
         int16_t prev_nold = 0;
         int pg_nfree = 0;
-        jl_taggedvalue_t **pfl_begin = NULL;
+        jl_value_t **pfl_begin = NULL;
         uint8_t msk = 1; // mask for the age bit in the current age byte
         while ((char*)v <= lim) {
             int bits = v->bits.gc;
+#ifdef MEMFENCE
+            // testing that the old header points either to value->type->datatype
+            // or that the it runs in a chain to hit NULL (e.g. it was in the old freelist)
+            jl_value_t *tv = jl_typeof(jl_valueof(v));
+            // 0x10 (type) and 0x20 (singleton) are used by the incremental serialize to invalidate objects
+            if (v->header != 0 && v->header != 0x10 && v->header != 0x20 && tv != (jl_value_t*)jl_buff_tag) {
+                jl_value_t *dt = jl_typeof(tv);
+                if (jl_astaggedvalue(tv)->header != 0) {
+                    assert(tv != (jl_value_t*)jl_buff_tag);
+                    jl_value_t *tdt = jl_typeof(dt);
+                    if (jl_astaggedvalue(dt)->header != 0 && tdt != (jl_value_t*)jl_datatype_type) {
+                        assert(page_metadata(dt) && page_metadata(tdt));
```

Comment:
yuyichao mentioned to me that this can check that the `osize` in the metadata here is the same as `osize`. also can check, if this point is reached, that the pointer chain of cells leading up to this point were unmarked

and can make sure it's pointing to the beginning of a cell (although this seems harder to compute, but see `jl_gc_find_taggedvalue_pool`)